File: Assemblies/Protean.Tools.vb/Object.vb
Imports System.Reflection
Imports Protean.Tools.Text

Public Class TypeExtensions

    Shared Function Type(ByVal typeFromCurrentAssembly As String) As System.Type

        Return System.Type.GetType(typeFromCurrentAssembly, True, True)

    End Function

    Shared Function Type(ByVal typeFromSpecifiedAssembly As String, ByVal assemblyName As String, Optional ByVal assemblyType As String = "") As System.Type

        If String.IsNullOrEmpty(assemblyName) Then
            ' Use the current assembly
            Return Type(typeFromSpecifiedAssembly)
        Else
            If Not String.IsNullOrEmpty(assemblyType) Then
                typeFromSpecifiedAssembly &= ", " & assemblyType
            End If
            Dim assemblyInstance As [Assembly] = [Assembly].Load(assemblyName)
            Return assemblyInstance.GetType(typeFromSpecifiedAssembly, True)
            assemblyInstance = Nothing
        End If

    End Function


    Public Class TypeMethodParser

        Private _type As String = ""
        Private _method As String = ""

        Public Sub New(ByVal typeAndMethod As String)
            Parse(typeAndMethod)
        End Sub

        Public ReadOnly Property TypeName() As String
            Get
                Return _type
            End Get
        End Property

        Public ReadOnly Property MethodName() As String
            Get
                Return _method
            End Get
        End Property

        Private Sub Parse(ByVal typeAndMethod As String)


            ' Searches for a fully qualified assembly name
            ' as specified in http://msdn.microsoft.com/en-us/library/w3f99sx1.aspx

            Dim typeMethodPattern As String = "(\w[\w\.\+\\]+)\.(\w+)"

            Dim typeName As String = SimpleRegexFind(typeAndMethod, typeMethodPattern, 1)
            Dim methodName As String = SimpleRegexFind(typeAndMethod, typeMethodPattern, 2)

            If String.IsNullOrEmpty(typeName) Or String.IsNullOrEmpty(methodName) Then

                Throw New FormatException("The input string did not match the expected format for type and methods.")

            Else
                _type = typeName
                _method = methodName
            End If

        End Sub

    End Class

End Class

File: Assemblies/Net-Standard-Protean.Tools/Object.cs
using System;
using System.Reflection;

namespace Protean.Tools
{
    public partial class TypeExtensions
    {

        public static Type Type(string typeFromCurrentAssembly)
        {

            return System.Type.GetType(typeFromCurrentAssembly, true, true);

        }

        public static Type Type(string typeFromSpecifiedAssembly, string assemblyName, string assemblyType = "")
        {

            if (string.IsNullOrEmpty(assemblyName))
            {
                // Use the current assembly
                return Type(typeFromSpecifiedAssembly);
            }
            else
            {
                if (!string.IsNullOrEmpty(assemblyType))
                {
                    typeFromSpecifiedAssembly += ", " + assemblyType;
                }
                var assemblyInstance = Assembly.Load(assemblyName);
                return assemblyInstance.GetType(typeFromSpecifiedAssembly, true);
                //assemblyInstance = null;
            }

        }


        public partial class TypeMethodParser
        {

            private string _type = "";
            private string _method = "";

            public TypeMethodParser(string typeAndMethod)
            {
                Parse(typeAndMethod);
            }

            public string TypeName
            {
                get
                {
                    return _type;
                }
            }

            public string MethodName
            {
                get
                {
                    return _method;
                }
            }

            private void Parse(string typeAndMethod)
            {


                // Searches for a fully qualified assembly name
                // as specified in http://msdn.microsoft.com/en-us/library/w3f99sx1.aspx

                string typeMethodPattern = @"(\w[\w\.\+\\]+)\.(\w+)";

                string typeName =Text.SimpleRegexFind(typeAndMethod, typeMethodPattern, 1);
                string methodName = Text.SimpleRegexFind(typeAndMethod, typeMethodPattern, 2);

                if (string.IsNullOrEmpty(typeName) | string.IsNullOrEmpty(methodName))
                {

                    throw new FormatException("The input string did not match the expected format for type and methods.");
                }

                else
                {
                    _type = typeName;
                    _method = methodName;
                }

            }

        }

    }
}
